### ISSUE TITLE
Bare CR allowed in code causing terrible spans

### ISSUE THREAD

**Issue by SiegeLord** [rust-lang/rust#16480]
E.g. consider this code (save it to a file with CR line endings):

``` rust
fn main() {
    let a = 0;
}
```

Error:

```
test.rs:1:18: 1:23 error: cannot determine a type for this local variable: cannot determine the type of this integer; add a suffix to specify the type explicitly [E0102]
}est.rs:let a = 0;) {
                            ^~~~~
error: aborting due to previous error
```

I'm pretty sure this did not used to happen.

I suggest disallowing bare CRs rather than fixing this span issue.

**Comment by steveklabnik** [comment]
Triage: no change.

**Comment by klensy** [comment]
Still the same:
`main.rs` with crlf or lf endings:
```rust
fn main() {
    let a = 0;
}
```
```
$ rustc main.rs
warning: unused variable: `a`
 --> main.rs:2:9
  |
2 |     let a = 0;
  |         ^ help: if this is intentional, prefix it with an underscore: `_a`
  |
  = note: `#[warn(unused_variables)]` on by default

warning: 1 warning emitted
```
same, main.rs, but with cr endings:
```
$ rustc main.rs
warning: unused variable: `a`
 --> main.rs:1:21
  |
}   let a = 0;{
  |                    ^ help: if this is intentional, prefix it with an underscore: `_a`
  |
  = note: `#[warn(unused_variables)]` on by default

warning: 1 warning emitted
```

**Comment by estebank** [comment]
Since https://github.com/rust-lang/rust/pull/90462 we have [replacement machinery for arbitrary codepoints](https://github.com/rust-lang/rust/pull/90462/files#diff-a35029c589a6b5b002ecf598efb24d174232e719ecbc6f741a58eee8509cd6a5R2066-R2078) that we could use for this. At the very least it would be easy to replace these with ` ` (if replacement with `\n` proves to require more extensive changes) to prevent further presentation issues quickly. At the same time, we should have a warn-by-default lint for these codepoints so that we can emit a targeted message about these codepoints with applicable suggestions.

**Comment by estebank** [comment]
This was fixed by #127528.
